### DISCUSSION TITLE
Component Testing in Monorepo. DevServer can't find webpack config

### DISCUSSION THREAD

**Discussion by bryanjos** [cypress-io/cypress#23289] [Questions and Help]
Hi 👋🏿 

Cypress: 10.4.0
@cypress/webpack-dev-server: 2.0.0
Next: 12.1.2

I'm migrating from Cypress 8.3.1 to 10.4.0. I've gotten most things working except for updates for component testing. I have a monorepo that's set up like so:

```
/packages
  /next
  /e2e <- where cypress and tests live
```

I'm updating the config from json to typescript and having trouble configuring the new devServer configuration to work.

Using the following, doesn't seem to work. I get a bunch of compilation errors from webpack because it can't find my configuration

```typescript
  component: {
    devServer: {
      framework: 'next',
      bundler: 'webpack',
    }
  }
```

Same with with using some code I had for 8.3.1 to get the webpack config and pass to the devServer

```typescript
    devServer(devServerConfig) {
     const webPackConfig = getWebpackConfig(devServerConfig.cypressConfig)

      return devServer({
        ...devServerConfig,
        framework: 'next',
        webPackConfig
      });
```

It feels like the webpackConfig is getting ignored.

I went digging into the `@cypress/webpack-dev-server` and looked at the `loadWebpackConfig` function in the `nextHandler` module. I see that it tries to use the `projectRoot` to try to get the next config in the call to `loadConfig`. If I manually change the second argument from `devServerConfig.cypressConfig.projectRoot` to the path to my next project, everything works. But of course that's not a viable solution to change the package like so.

So I was wondering if there is a better way to tell the devServer where my next config lives?

**Comment by portyu9** [comment]
Hi @bryanjos 

The problem is that Cypress's `projectRoot` is your `/e2e` package, but the actual Next.js project is `/packages/next`.

For Next component testing, make the **Next application directory the Cypress project root**:

```bash
npx cypress open --component --project packages/next
```

and keep the component-testing configuration associated with that Next project:

```js
component: {
  devServer: {
    framework: 'next',
    bundler: 'webpack',
  },
}
```

Cypress then resolves the correct Next configuration instead of looking under the separate `e2e` package.

Do not patch `@cypress/webpack-dev-server`'s `projectRoot`; run component testing from the application it is actually compiling.

Please mark this answer as accepted if it helped.
